tiledb/cloud/notebook.py

```diff
@@ -35,3 +50,290 @@ def rename_notebook(
+    :param storage_path: such as "s3://acmecorp-janedoe", typically from the
+      user's account settings.
+    :param array_name : name to be seen in the UI, such as "testing-upload"
+    :param namespace: such as "janedoe".
+    :param storage_credential_name: such as "janedoe-creds", typically from the
+      user's account settings.
+    :return: TileDB array name, such as "tiledb://janedoe/testing-upload".
+    """
+
+    ctx = tiledb.cloud.Ctx(
+        {
+            "rest.creation_access_credentials_name": storage_credential_name,
+        }
+    )
+
+    tiledb_uri, array_name = _create_notebook_array(
+        storage_path,
+        array_name,
+        namespace,
+        ctx,
+    )
+
+    _write_notebook_to_array(tiledb_uri, ipynb_file_contents, ctx)
+
+    return tiledb_uri
+
+
+def _create_notebook_array(
+    storage_path: str,
+    array_name: str,
+    namespace: str,
+    ctx: tiledb.Ctx,
+    *,
+    retries: int = 0,
+) -> Tuple[str, str]:
+    """
+    Creates a new array for storing a notebook file.
+    :param storage_path: such as "s3://acmecorp-janedoe", typically from the
+      user's account settings.
+    :param array_name : name to be seen in the UI, such as "testing-upload"
+    :param namespace: such as "janedoe".
+    :param ctx: cloud context for the operation.
+    :return: tuple of tiledb_uri and array_name
+    """
+
+    if namespace in RESERVED_NAMESPACES:
+        raise ValueError(
+            f"{namespace!r} is not a valid folder to create notebooks. "
+            "Please select a proper namespace (username or organization name).",
+        )
+
+    # The array will be be 1-dimensional with domain of 0 to max uint64. We
+    # use a tile extent of 1024 bytes.
+    dom = tiledb.Domain(
+        tiledb.Dim(
+            name="position",
+            domain=(0, numpy.iinfo(numpy.uint64).max - 1025),
+            tile=1024,
+            dtype=numpy.uint64,
+            ctx=ctx,
+        ),
+        ctx=ctx,
+    )
+
+    tries = 1 + retries  # 1st + rest
+    while tries > 0:
+        try:
+            tiledb_uri, array_name = _create_notebook_array_retry_helper(
+                storage_path,
+                array_name,
+                namespace,
+                dom,
+                ctx,
+            )
+            return (tiledb_uri, array_name)
+        except tiledb.TileDBError as e:
+            if "Error while listing with prefix" in str(e):
+                # It is possible to land here if user sets wrong default S3
+                # credentials with respect to default S3 path.
+                raise ValueError(
+                    f"Error creating file: {e}. Are your credentials valid?"
+                ) from e
+            if "Cannot create array" in str(e) and "already exists" in str(e):
+                raise ValueError(
+                    f"Error creating file: {array_name} already exists in namespace {namespace}."
+                )
+        # Retry other exceptions
```

Comment:
I would bump this in by four spaces so that it is in the except block and note that it is "retry other TileDBErrors", because other exception types will still bubble up.